cpu/esp32/irq_arch.c

```diff
@@ -19,58 +19,101 @@
  */
 
 #include "irq_arch.h"
+
+#include "esp_attr.h"
 #include "esp_err.h"
-#include "esp32/rom/ets_sys.h"
-#include "soc/dport_reg.h"
-#include "xtensa/xtensa_api.h"
+#include "freertos/FreeRTOS.h"
+#include "hal/interrupt_controller_types.h"
+#include "hal/interrupt_controller_ll.h"
+#include "rom/ets_sys.h"
+#include "soc/periph_defs.h"
+#include "esp_intr_alloc.h"
 
 #define ENABLE_DEBUG 0
 #include "debug.h"
 
-struct _irq_alloc_table_t {
+#ifndef ETS_CAN_INTR_SOURCE
+#define ETS_CAN_INTR_SOURCE     ETS_TWAI_INTR_SOURCE
+#endif
+
+typedef struct intr_handle_data_t {
     int      src;   /* peripheral interrupt source */
     uint32_t intr;  /* interrupt number */
+    uint8_t  level;
+} intr_handle_data_t;
+
+/* TODO change to a clearer approach */
+static struct intr_handle_data_t _irq_data_table[] = {
```

Comment:
Can this be `const`?

Reply:
Unfortunatly not. The address of the according element in the table is used as handle and the assignment here https://github.com/RIOT-OS/RIOT/blob/7fbf42ebb869aacf582c864962054a3baf406c4b/cpu/esp32/irq_arch.c#L161 discards the const qualifier.

The index in the table could also be used as a handle. But I decided to spend these bytes in RAM for the table and to use the address of the element as handle instead of the index, because this can then be used directly in all other functions that get the handle as parameter. It was just a tradeoff of performance and ressources. The size of the RAM is of secondary importance in ESP SoCs.

Reply:
I pushed commit 03d1b16822e8ca9ae57b97c21b67aa23e6361836 just for demonstration how it could be implemented with the index as interrupt handle and with a number of type casts.

Unfortunatly, the ESP-IDF uses a forward declaration for `struct intr_handle_data_t` and declares `intr_handle_t` as pointer to this structure.
```
typedef struct intr_handle_data_t intr_handle_data_t;
typedef intr_handle_data_t *intr_handle_t ;
```
That is, it expects that `intr_handle_t` is a pointer to a `struct intr_handle_data_t` without defining its structure. That's why, I used the structure of entries in the interrupt data table directly as `struct intr_handle_data_t` and the address of the elements as `intr_handle_t`.

With commit 036b1d66e380e0c13bb4462adae86e98ceb7f682, a simple type cast would allow to declare the table as static const which saves 192 byte (with `uint32_t` for the interrupt number) or 128 bytes (with `uint8_t` for the interrupt number). IMHO, this would be a good compromise and more clear than the tricky use of index as interrupt handle.

Reply:
@benpicco Which one of the two approaches would you prefer? I would clean up the PR for the next compilation run accordingly.

BTW, the hashes seem to be still a problem.

Reply:
I think you can better judge the trade-offs of the two solutions, I'm fine with either.

Unfortunately I don't have any ideas what causes the hash mismatch.

Reply:
> Unfortunately I don't have any ideas what causes the hash mismatch.

Ah, this time I had luck. The compilation has been successful :+1:

Reply:
> Just squash if you have something you are happy with.

Finally I squashed that version which uses the address of the elements as handle and the single type cast.